python/pyiceberg/io/pyarrow.py

```diff
@@ -476,6 +483,202 @@ def expression_to_pyarrow(expr: BooleanExpression) -> pc.Expression:
+    def schema(self, schema: pa.Schema, field_results: List[IcebergType]) -> Schema:
+        fields = []
+        for i in range(len(schema.names)):
+            field = schema.field(i)
+            field_id = _get_field_id(field)
+            field_type = field_results[i]
+            if field_type is not None:
+                fields.append(NestedField(field_id, field.name, field_type, required=not field.nullable))
+        return Schema(*fields)
+
+    def struct(self, struct: pa.StructType, field_results: List[IcebergType]) -> IcebergType:
+        fields = []
+        for i in range(struct.num_fields):
+            field = struct[i]
+            field_id = _get_field_id(field)
+            # may need to check doc strings
+            field_type = field_results[i]
+            if field_type is not None:
+                fields.append(NestedField(field_id, field.name, field_type, required=not field.nullable))
+        return StructType(*fields)
+
+    def list(self, list_type: pa.ListType, element_result: IcebergType) -> IcebergType:
+        element_field = list_type.value_field
+        element_id = _get_field_id(element_field)
+        if element_result is not None:
+            return ListType(element_id, element_result, element_required=not element_field.nullable)
+        raise ValueError(f"List type must have element field: {list_type}")
+
+    def map(self, map_type: pa.MapType, key_result: IcebergType, value_result: IcebergType) -> IcebergType:
+        key_field = map_type.key_field
+        key_id = _get_field_id(key_field)
+        value_field = map_type.item_field
+        value_id = _get_field_id(value_field)
+        if key_result is not None and value_result is not None:
+            return MapType(key_id, key_result, value_id, value_result, value_required=not value_field.nullable)
+        raise ValueError(f"Map type must have key and value fields: {map_type}")
+
+    def primitive(self, primitive: pa.DataType) -> IcebergType:
+        if pa.types.is_boolean(primitive):
+            return BooleanType()
+        elif pa.types.is_int32(primitive) or pa.types.is_uint32(primitive):
+            return IntegerType()
+        elif pa.types.is_int64(primitive) or pa.types.is_uint64(primitive):
+            return LongType()
+        elif pa.types.is_float32(primitive):
+            return FloatType()
+        elif pa.types.is_float64(primitive):
+            return DoubleType()
+        elif pa.types.is_decimal(primitive):
+            if isinstance(primitive, pa.Decimal256Type):
+                primitive = cast(pa.Decimal256Type, primitive)
+            else:
+                primitive = cast(pa.Decimal128Type, primitive)
+            return DecimalType(primitive.precision, primitive.scale)
+        elif pa.types.is_string(primitive):
+            return StringType()
+        elif pa.types.is_date(primitive):
+            return DateType()
+        elif pa.types.is_time(primitive):
```

Comment:
Do we need to check time or timestamp precision?

Reply:
@Fokko, I think this could be a problem. We should make sure there are good tests for these types.

The read code operates using PyIceberg schemas. So that is going to convert PyArrow to PyIceberg, prune columns, and the convert back to request a projection from PyArrow. If there is a lossy conversion then that will cause a problem. For example if a time type is in millis but is converted without that information, then the requested type will be a timestamp in micros and that may cause a problem.

When looking at this, I found another problem. Rather than requesting a specific schema, the projected Iceberg schema is used to request top-level columns:

```python
        arrow_table = pq.read_table(
            source=fout,
            schema=parquet_schema,
            pre_buffer=True,
            buffer_size=8 * ONE_MEGABYTE,
            filters=pyarrow_filter,
            columns=[col.name for col in file_project_schema.columns],
        )
```

I assume that's going to read full nested structures because the only information it has is the top-level column name. If that happens, I'm concerned that we're reading leaf columns that we don't actually need to. I think projection will correctly ignore them, but it would be nice to avoid reading them.

Reply:
> The read code operates using PyIceberg schemas. So that is going to convert PyArrow to PyIceberg, prune columns, and the convert back to request a projection from PyArrow. If there is a lossy conversion then that will cause a problem. For example if a time type is in millis but is converted without that information, then the requested type will be a timestamp in micros and that may cause a problem.

Good point, I think we should have round-trip tests here, where we go from Iceberg to PyArrow, and back to Iceberg.

> I assume that's going to read full nested structures because the only information it has is the top-level column name. If that happens, I'm concerned that we're reading leaf columns that we don't actually need to. I think projection will correctly ignore them, but it would be nice to avoid reading them.

I agree, I'm would have to look into the details if we prune the fields of the nested fields properly. In the future I'd like to replace this with the custom evolution strategy: https://github.com/apache/arrow/issues/33972#issuecomment-1414589158

Reply:
Thank you for your explanation. I refactored the code to make it consistent with the primitive conversion in iceberg to pyarrow conversion. Currently, It only allow UTC as timezone and us as unit.

I also added some round trip tests for these types

Reply:
This still seems suspicious to me. While it's now correct, how does PyArrow read Python files that have other time or timestamp representations?

Reply:
Thank you for your review and for bringing up your concerns. I'd like to understand better what you find suspicious about PyArrow's ability to read Python files with different time or timestamp representations.

From what I understand, Iceberg's TimeType, TimestampType, and TimestampzType require `us` and `UTC`, and my current conversion ensures that no data is lost.

In this case, PyArrow can support reading non-UTC timezones and `s`, `ms`, and `us` precision, but it does not support nanosecond precision since the final requested type during projection will be `us` and `UTC`.: https://github.com/apache/iceberg/blob/283107d3608e65804314ba8fd2cb76aafdc36bbd/python/pyiceberg/io/pyarrow.py#L397-L404

I chose to restrict the precision to `us` and the timezone to `UTC` because the Iceberg specification requires all stored time/timestamp to be in this precision and timezone. Since the `pyarrow_to_schema` visitor is used to read an Iceberg table's data file, I believe we should only support `us` and `UTC`.

However, I am also not very sure about it. Regarding support for other precision and timezone here, I think more discussion and modifications may be needed if we want to add other support. How about creating another PR if needed to address these concerns?

Thank you again for your feedback, and please let me know if you have any further questions or concerns.

Reply:
If there is an Iceberg schema in the file, I think we can assume that it is written according to the spec:
![image](https://user-images.githubusercontent.com/1134248/234511353-ce691d90-4341-4369-a7ca-41f3bb33c8b7.png)

With the current check, it is correct:
```python
        elif pa.types.is_time(primitive):
            if isinstance(primitive, pa.Time64Type) and primitive.unit == "us":
                return TimeType()
```

We could even simplify it:

```python
        elif isinstance(primitive, pa.Time64Type) and primitive.unit == "us":
            return TimeType()
```

The tests are in place:

```python
def test_pyarrow_time32_to_iceberg() -> None:
    pyarrow_type = pa.time32("ms")
    with pytest.raises(TypeError, match=re.escape("Unsupported type: time32[ms]")):
        visit_pyarrow(pyarrow_type, _ConvertToIceberg())
    pyarrow_type = pa.time32("s")
    with pytest.raises(TypeError, match=re.escape("Unsupported type: time32[s]")):
        visit_pyarrow(pyarrow_type, _ConvertToIceberg())


def test_pyarrow_time64_us_to_iceberg() -> None:
    pyarrow_type = pa.time64("us")
    converted_iceberg_type = visit_pyarrow(pyarrow_type, _ConvertToIceberg())
    assert converted_iceberg_type == TimeType()
    assert visit(converted_iceberg_type, _ConvertToArrowSchema()) == pyarrow_type


def test_pyarrow_time64_ns_to_iceberg() -> None:
    pyarrow_type = pa.time64("ns")
    with pytest.raises(TypeError, match=re.escape("Unsupported type: time64[ns]")):
        visit_pyarrow(pyarrow_type, _ConvertToIceberg())
```

I think we can resolve this issue